### ISSUE TITLE
Class 'phpseclib\Crypt\RSA' not found

### ISSUE THREAD

**Issue by alve89** [phpseclib/phpseclib#1459]
Hey guys,

after installing this version: https://github.com/phpseclib/phpseclib/releases/tag/2.0.25
and the latest version of composer I wanted to use this code

```
<?php
include('/www/htdocs/XXXXX/vendor/autoload.php');

 $rsa = new \phpseclib\Crypt\RSA();
 extract($rsa->createKey());

 $plaintext = 'terrafrost';

 $rsa->loadKey($privatekey);
 $signature = $rsa->sign($plaintext);

 $rsa->loadKey($publickey);
 echo $rsa->verify($plaintext, $signature) ? 'verified' : 'unverified';
```

but I am getting the following error (that's the only error):

> Fatal error: Uncaught Error: Class 'phpseclib\Crypt\RSA' not found in /www/htdocs/XXXXX/keygen.php:10 Stack trace: #0 {main} thrown in /www/htdocs/XXXXX/keygen.php on line 10

Could someone please help me here?

Thanks and regards
Stefan

**Comment by terrafrost** [comment]
I wonder what would happen if you did `$rsa = new \phpseclib3\Crypt\RSA();`. The 3.0 and master branches use a different namespace then the 2.0 branch does.

**Comment by alve89** [comment]
It's actually the same: 

> Fatal error: Uncaught Error: Class 'phpseclib3\Crypt\RSA' not found in /www/htdocs/XXXXX/keygen.php:11 Stack trace: #0 {main} thrown in /www/htdocs/XXXXXkeygen.php on line 11

After cloning the 3.0 branch I'm getting this:

> Fatal error: Uncaught Error: Cannot instantiate abstract class phpseclib3\Crypt\RSA in /www/htdocs/XXXXX/keygen.php:11 Stack trace: #0 {main} thrown in /www/htdocs/w013e4e2/XXXXX/keygen.php on line 11

**Comment by terrafrost** [comment]
> After cloning the 3.0 branch I'm getting this:
>
>> Fatal error: Uncaught Error: Cannot instantiate abstract class phpseclib3\Crypt\RSA in /www/htdocs/XXXXX/keygen.php:11 Stack trace: #0 {main} thrown in /www/htdocs/w013e4e2/XXXXX/keygen.php on line 11

Well at least that's working correctly.

Can you post your composer.json and your composer.lock files? Can you go into your vendor directory and see if there's a phpseclib directory? If there is can you send me a list of files in that directory?

Thanks!

**Comment by alve89** [comment]
**composer.json**
```
{
    "name": "phpseclib/phpseclib",
    "type": "library",
    "description": "PHP Secure Communications Library - Pure-PHP implementations of RSA, AES, SSH2, SFTP, X.509 etc.",
    "keywords": [
        "security",
        "crypto",
        "cryptography",
        "encryption",
        "signature",
        "signing",
        "rsa",
        "aes",
        "blowfish",
        "twofish",
        "ssh",
        "sftp",
        "x509",
        "x.509",
        "asn1",
        "asn.1",
        "BigInteger"
        ],
    "homepage": "http://phpseclib.sourceforge.net",
    "license": "MIT",
    "authors": [
        {
            "name": "Jim Wigginton",
            "email": "terrafrost@php.net",
            "role": "Lead Developer"
        },
        {
            "name": "Patrick Monnerat",
            "email": "pm@datasphere.ch",
            "role": "Developer"
        },
        {
            "name": "Andreas Fischer",
            "email": "bantu@phpbb.com",
            "role": "Developer"
        },
        {
            "name": "Hans-Jürgen Petrich",
            "email": "petrich@tronic-media.com",
            "role": "Developer"
        },
        {
            "name": "Graham Campbell",
            "email": "graham@alt-three.com",
            "role": "Developer"
        }
    ],
    "require": {
        "paragonie/constant_time_encoding": "^1",
        "paragonie/random_compat": "^1.4|^2.0",
        "php": ">=5.6.1"
    },
    "require-dev": {
        "phing/phing": "~2.7",
        "phpunit/phpunit": "^4.8.35|^5.7|^6.0",
        "sami/sami": "~2.0",
        "squizlabs/php_codesniffer": "~2.0"
    },
    "suggest": {
        "ext-libsodium": "SSH2/SFTP can make use of some algorithms provided by the libsodium-php extension.",
        "ext-openssl": "Install the OpenSSL extension in order to speed up a wide variety of cryptographic operations.",
        "ext-mcrypt": "Install the Mcrypt extension in order to speed up a few other cryptographic operations.",
        "ext-gmp": "Install the GMP (GNU Multiple Precision) extension in order to speed up arbitrary precision integer arithmetic operations."
    },
    "autoload": {
        "files": [
            "phpseclib/bootstrap.php"
        ],
        "psr-4": {
            "phpseclib3\\": "phpseclib/"
        }
    }
}

```
**composer.lock**
I can't find this file...?

> Can you go into your vendor directory and see if there's a phpseclib directory? If there is can you send me a list of files in that directory?
There's no phpseclib directory.

**Comment by terrafrost** [comment]
You posted phpseclib's composer.json but not your projects composer.json.

What does your projects root directory look like? Can you do `ls -latr` or `dir` or whatever is appropriate for your OS and post it? Also, while we're at it, go ahead and do the same thing for your vendor/ directory.

**Comment by alve89** [comment]
> You posted phpseclib's composer.json but not your projects composer.json.

I don't have a "project". But I'm going to write a tool for myself to generate certificates and use them in other applications (like my [PHPmobileConfig](https://github.com/alve89/PHPMobileConfig) ). But first I need to get your library / tool work.


> What does your projects root directory look like? Can you do `ls -latr` or `dir` or whatever is appropriate for your OS and post it?

It only has one file yet: keygen.php (its content is only the code I posted above). I just started to try your lib.
![image](https://user-images.githubusercontent.com/11487054/76286321-1a148d80-62a2-11ea-855b-f4e926c575c1.png)

> Also, while we're at it, go ahead and do the same thing for your vendor/ directory.

![image](https://user-images.githubusercontent.com/11487054/76285779-d9684480-62a0-11ea-8cd3-a2ff2d483a0a.png)

**Comment by terrafrost** [comment]
You'll need to do `composer require phpseclib/phpseclib:~2.0` in your "projects" root directory.

To do that you'll first need to first download Composer: https://getcomposer.org/download/ Then you'll need to install it: https://getcomposer.org/doc/00-intro.md#installation-linux-unix-macos I always install it locally.

**Comment by alve89** [comment]
That worked..... I never used the composer before, so I just downloaded your lib and copied the composer.json to my projects root, where it was found by the composer (at least I think that). So: Sorry and thank you for your time!

Simple last question: I tried that (http://phpseclib.sourceforge.net/rsa/2.0/examples.html) to generate a key-pair - where are the keys stored / how can I get the content of the keys?
